File: src/engine/gates.ts
import { Signal, SignalState, GateType, ComponentType } from '@/types/circuit';

function resolve(signal: SignalState): Signal {
  return signal ?? 0;
}

function evaluateAND(inputs: SignalState[]): Signal {
  if (inputs.length === 0) return 0;
  return inputs.every((s) => resolve(s) === 1) ? 1 : 0;
}

function evaluateOR(inputs: SignalState[]): Signal {
  if (inputs.length === 0) return 0;
  return inputs.some((s) => resolve(s) === 1) ? 1 : 0;
}

function evaluateNOT(inputs: SignalState[]): Signal {
  if (inputs.length === 0) return 1;
  return resolve(inputs[0]) === 1 ? 0 : 1;
}

function evaluateNAND(inputs: SignalState[]): Signal {
  if (inputs.length === 0) return 1;
  return inputs.every((s) => resolve(s) === 1) ? 0 : 1;
}

function evaluateNOR(inputs: SignalState[]): Signal {
  if (inputs.length === 0) return 1;
  return inputs.some((s) => resolve(s) === 1) ? 0 : 1;
}

function evaluateXOR(inputs: SignalState[]): Signal {
  if (inputs.length === 0) return 0;
  const xorResult = inputs.reduce<number>((acc, s) => acc ^ resolve(s), 0);
  return (xorResult & 1) as Signal;
}

function evaluateXNOR(inputs: SignalState[]): Signal {
  if (inputs.length === 0) return 1;
  const xorResult = inputs.reduce<number>((acc, s) => acc ^ resolve(s), 0);
  return ((xorResult & 1) === 0 ? 1 : 0) as Signal;
}

function evaluateBUFFER(inputs: SignalState[]): Signal {
  if (inputs.length === 0) return 0;
  return resolve(inputs[0]);
}

const GATE_EVALUATORS: Record<GateType, (inputs: SignalState[]) => Signal> = {
  AND: evaluateAND,
  OR: evaluateOR,
  NOT: evaluateNOT,
  NAND: evaluateNAND,
  NOR: evaluateNOR,
  XOR: evaluateXOR,
  XNOR: evaluateXNOR,
  BUFFER: evaluateBUFFER,
};

export function evaluateGate(type: GateType, inputs: SignalState[]): Signal {
  const evaluator = GATE_EVALUATORS[type];
  if (!evaluator) {
    console.warn(`Unknown gate type: ${type}, returning 0`);
    return 0;
  }
  return evaluator(inputs);
}

export function isGateType(type: ComponentType): type is GateType {
  return type in GATE_EVALUATORS;
}


export function evaluateFullAdder(
  a: SignalState,
  b: SignalState,
  cin: SignalState
): { sum: Signal; cout: Signal } {
  const ra = resolve(a);
  const rb = resolve(b);
  const rc = resolve(cin);
  const axb = (ra ^ rb) & 1;
  return {
    sum: ((axb ^ rc) & 1) as Signal,
    cout: (((ra & rb) | (rc & axb)) & 1) as Signal,
  };
}

export function evaluateDecoder(
  a: SignalState,
  b: SignalState
): Signal[] {
  const ra = resolve(a);
  const rb = resolve(b);
  const index = (rb << 1) | ra;
  return [0, 1, 2, 3].map((i) => (i === index ? 1 : 0)) as Signal[];
}

export function evaluateComponent(
  type: ComponentType,
  inputs: Map<string, SignalState>,
  properties?: Record<string, unknown>
): Map<string, Signal> {
  const outputs = new Map<string, Signal>();

  switch (type) {
    case 'INPUT': {
      const value = inputs.get('value');
      outputs.set('out', resolve(value));
      break;
    }
    case 'OUTPUT':
    case 'LED': {
      const displaySignal = inputs.get('in_0') ?? Array.from(inputs.values())[0];
      outputs.set('display', resolve(displaySignal));
      break;
    }
    case 'CLOCK': {
      const value = inputs.get('value');
      outputs.set('out', resolve(value));
      break;
    }
    case 'SEVEN_SEGMENT': {
      const inputArr = Array.from(inputs.values());
      for (let i = 0; i < inputArr.length; i++) {
        outputs.set(`in_${i}`, resolve(inputArr[i]));
      }
      break;
    }
    case 'HALF_ADDER': {
      const a = resolve(inputs.get('in_0'));
      const b = resolve(inputs.get('in_1'));
      outputs.set('sum', (a ^ b) as Signal);
      outputs.set('carry', (a & b) as Signal);
      break;
    }
    case 'FULL_ADDER': {
      const a = resolve(inputs.get('in_0'));
      const b = resolve(inputs.get('in_1'));
      const cin = resolve(inputs.get('in_2'));
      const res = evaluateFullAdder(a, b, cin);
      outputs.set('sum', res.sum);
      outputs.set('cout', res.cout);
      break;
    }
    case 'DECODER': {
      const a = resolve(inputs.get('in_0'));
      const b = resolve(inputs.get('in_1'));
      const res = evaluateDecoder(a, b);
      // Pin IDs must match createOutputPins: y0, y1, y2, y3
      res.forEach((val, i) => outputs.set(`y${i}`, val));
      break;
    }

    // ── BCD TO 7-SEGMENT DECODER ──────────────────────
    // Inputs: D, C, B, A  →  Outputs: A, B, C, D, E, F, G
    case 'BCD_TO_7SEG': {
      const d = resolve(inputs.get('in_0'));
      const c = resolve(inputs.get('in_1'));
      const b = resolve(inputs.get('in_2'));
      const a = resolve(inputs.get('in_3'));
      const val = (d << 3) | (c << 2) | (b << 1) | a;

      // Hexadecimal 7-segment active-high patterns for 0-F (A=MSB..G=LSB order in logic below is typical, but we map 0=A..6=G)
      // Array stores [A, B, C, D, E, F, G] for each number
      const patterns = [
        [1,1,1,1,1,1,0], // 0
        [0,1,1,0,0,0,0], // 1
        [1,1,0,1,1,0,1], // 2
        [1,1,1,1,0,0,1], // 3
        [0,1,1,0,0,1,1], // 4
        [1,0,1,1,0,1,1], // 5
        [1,0,1,1,1,1,1], // 6
        [1,1,1,0,0,0,0], // 7
        [1,1,1,1,1,1,1], // 8
        [1,1,1,1,0,1,1], // 9
        [1,1,1,0,1,1,1], // A
        [0,0,1,1,1,1,1], // b
        [1,0,0,1,1,1,0], // C
        [0,1,1,1,1,0,1], // d
        [1,0,0,1,1,1,1], // E
        [1,0,0,0,1,1,1], // F
      ];
      const pattern = patterns[val] || [0,0,0,0,0,0,0];
      for (let i = 0; i < 7; i++) {
        outputs.set(`out_${i}`, pattern[i] as Signal);
      }
      break;
    }

    // ── MUX 2:1 ──────────────────────────────────────
    // Inputs: D0, D1, S  →  Output: Y
    // Y = S ? D1 : D0
    case 'MUX_2TO1': {
      const d0 = resolve(inputs.get('in_0'));
      const d1 = resolve(inputs.get('in_1'));
      const s  = resolve(inputs.get('in_2'));
      outputs.set('out', (s === 1 ? d1 : d0) as Signal);
      break;
    }

    // ── MUX 4:1 ──────────────────────────────────────
    // Inputs: D0, D1, D2, D3, S0, S1  →  Output: Y
    case 'MUX_4TO1': {
      const d0 = resolve(inputs.get('in_0'));
      const d1 = resolve(inputs.get('in_1'));
      const d2 = resolve(inputs.get('in_2'));
      const d3 = resolve(inputs.get('in_3'));
      const s0 = resolve(inputs.get('in_4'));
      const s1 = resolve(inputs.get('in_5'));
      const sel = (s1 << 1) | s0;
      const muxData = [d0, d1, d2, d3];
      outputs.set('out', muxData[sel] as Signal);
      break;
    }

    // ── DEMUX 1:4 ─────────────────────────────────────
    // Inputs: D, S0, S1  →  Outputs: Y0, Y1, Y2, Y3
    case 'DEMUX_1TO4': {
      const d  = resolve(inputs.get('in_0'));
      const s0 = resolve(inputs.get('in_1'));
      const s1 = resolve(inputs.get('in_2'));
      const sel = (s1 << 1) | s0;
      for (let i = 0; i < 4; i++) {
        outputs.set(`y${i}`, (i === sel ? d : 0) as Signal);
      }
      break;
    }

    // ── SR LATCH (level-sensitive, active-high) ──────
    // Inputs: S, R  →  Outputs: Q, Q̄
    case 'SR_LATCH': {
      const s = resolve(inputs.get('in_0'));
      const r = resolve(inputs.get('in_1'));
      const prevQ = (properties?.q as Signal) ?? 0;
      let q: Signal = prevQ;
      if (s === 1 && r === 0) q = 1;
      else if (s === 0 && r === 1) q = 0;
      else if (s === 1 && r === 1) q = 0; // invalid → reset
      // s === 0 && r === 0: hold
      if (properties) properties.q = q;
      outputs.set('q', q);
      outputs.set('qn', (q === 1 ? 0 : 1) as Signal);
      break;
    }

    // ── D FLIP-FLOP (positive-edge-triggered) ────────
    // Inputs: D, CLK  →  Outputs: Q, Q̄
    case 'D_FLIPFLOP': {
      const d   = resolve(inputs.get('in_0'));
      const clk = resolve(inputs.get('in_1'));
      const prevClk = (properties?.prevClk as Signal) ?? 0;
      let q = (properties?.q as Signal) ?? 0;
      let sampledD = (properties?.sampledD as Signal) ?? d;

      // Master-slave emulation: sample inputs while clock is low to avoid zero-delay race conditions
      if (clk === 0) {
        sampledD = d;
      }

      // Rising edge: prevClk=0, clk=1
      if (prevClk === 0 && clk === 1) {
        q = sampledD;
      }
      if (properties) {
        properties.prevClk = clk;
        properties.sampledD = sampledD;
        properties.q = q;
      }
      outputs.set('q', q);
      outputs.set('qn', (q === 1 ? 0 : 1) as Signal);
      break;
    }

    // ── JK FLIP-FLOP (positive-edge-triggered) ───────
    // Inputs: J, K, CLK  →  Outputs: Q, Q̄
    case 'JK_FLIPFLOP': {
      const j   = resolve(inputs.get('in_0'));
      const k   = resolve(inputs.get('in_1'));
      const clk = resolve(inputs.get('in_2'));
      const prevClk = (properties?.prevClk as Signal) ?? 0;
      let q = (properties?.q as Signal) ?? 0;
      let sampledJ = (properties?.sampledJ as Signal) ?? j;
      let sampledK = (properties?.sampledK as Signal) ?? k;

      // Master-slave emulation: sample inputs while clock is low
      if (clk === 0) {
        sampledJ = j;
        sampledK = k;
      }

      // Rising edge trigger using sampled values
      if (prevClk === 0 && clk === 1) {
        if (sampledJ === 0 && sampledK === 0) { /* hold */ }
        else if (sampledJ === 0 && sampledK === 1) q = 0;
        else if (sampledJ === 1 && sampledK === 0) q = 1;
        else q = (q === 1 ? 0 : 1) as Signal; // toggle
      }
      if (properties) {
        properties.prevClk = clk;
        properties.sampledJ = sampledJ;
        properties.sampledK = sampledK;
        properties.q = q;
      }
      outputs.set('q', q);
      outputs.set('qn', (q === 1 ? 0 : 1) as Signal);
      break;
    }

    // ── 1-BIT COMPARATOR ─────────────────────────────
    // Inputs: A, B  →  Outputs: GT (A>B), EQ (A=B), LT (A<B)
    case 'COMPARATOR': {
      const a = resolve(inputs.get('in_0'));
      const b = resolve(inputs.get('in_1'));
      outputs.set('gt', (a > b ? 1 : 0) as Signal);
      outputs.set('eq', (a === b ? 1 : 0) as Signal);
      outputs.set('lt', (a < b ? 1 : 0) as Signal);
      break;
    }

    case 'IC': {
      // IC logic is handled by the SimulationEngine directly
      // using sub-engines. We return an empty map here.
      break;
    }
    default: {
      if (isGateType(type)) {
        const inputArr = Array.from(inputs.values());
        const result = evaluateGate(type, inputArr);
        outputs.set('out', result);
      }
    }
  }

  return outputs;
}

export { GATE_EVALUATORS };

File: engine-wasm/src/gates.rs
use crate::types::{ComponentType, Signal};
use std::collections::HashMap;

fn resolve(s: Signal) -> Signal {
    s.resolve()
}

fn r(s: Signal) -> u8 {
    resolve(s).as_bit()
}

// ── Gate evaluators ──────────────────────────────────────────

pub fn evaluate_gate(gate_type: ComponentType, inputs: &[Signal]) -> Signal {
    match gate_type {
        ComponentType::And => eval_and(inputs),
        ComponentType::Or => eval_or(inputs),
        ComponentType::Not => eval_not(inputs),
        ComponentType::Nand => eval_nand(inputs),
        ComponentType::Nor => eval_nor(inputs),
        ComponentType::Xor => eval_xor(inputs),
        ComponentType::Xnor => eval_xnor(inputs),
        ComponentType::Buffer => eval_buffer(inputs),
        _ => Signal::Low,
    }
}

fn eval_and(inputs: &[Signal]) -> Signal {
    if inputs.is_empty() {
        return Signal::Low;
    }
    Signal::from_bit(inputs.iter().all(|s| r(*s) == 1) as u8)
}

fn eval_or(inputs: &[Signal]) -> Signal {
    if inputs.is_empty() {
        return Signal::Low;
    }
    Signal::from_bit(inputs.iter().any(|s| r(*s) == 1) as u8)
}

fn eval_not(inputs: &[Signal]) -> Signal {
    if inputs.is_empty() {
        return Signal::High;
    }
    Signal::from_bit(if r(inputs[0]) == 1 { 0 } else { 1 })
}

fn eval_nand(inputs: &[Signal]) -> Signal {
    if inputs.is_empty() {
        return Signal::High;
    }
    Signal::from_bit(!inputs.iter().all(|s| r(*s) == 1) as u8)
}

fn eval_nor(inputs: &[Signal]) -> Signal {
    if inputs.is_empty() {
        return Signal::High;
    }
    Signal::from_bit(!inputs.iter().any(|s| r(*s) == 1) as u8)
}

fn eval_xor(inputs: &[Signal]) -> Signal {
    if inputs.is_empty() {
        return Signal::Low;
    }
    let xor = inputs.iter().fold(0u8, |acc, s| acc ^ r(*s));
    Signal::from_bit(xor & 1)
}

fn eval_xnor(inputs: &[Signal]) -> Signal {
    if inputs.is_empty() {
        return Signal::High;
    }
    let xor = inputs.iter().fold(0u8, |acc, s| acc ^ r(*s));
    Signal::from_bit(if (xor & 1) == 0 { 1 } else { 0 })
}

fn eval_buffer(inputs: &[Signal]) -> Signal {
    if inputs.is_empty() {
        return Signal::Low;
    }
    resolve(inputs[0])
}

// ── Component evaluators ─────────────────────────────────────

pub fn evaluate_component(
    comp_type: ComponentType,
    inputs: &HashMap<String, Signal>,
    properties: &mut HashMap<String, serde_json::Value>,
) -> HashMap<String, Signal> {
    let mut outputs = HashMap::new();

    match comp_type {
        ComponentType::Input => {
            let value = properties
                .get("value")
                .and_then(|v| v.as_u64())
                .unwrap_or(0);
            outputs.insert("out".to_string(), Signal::from_bit(value as u8));
        }

        ComponentType::Output | ComponentType::Led => {
            let display = inputs
                .get("in_0")
                .copied()
                .or_else(|| inputs.values().next().copied())
                .unwrap_or(Signal::Low);
            outputs.insert("display".to_string(), resolve(display));
        }

        ComponentType::Clock => {
            let value = properties
                .get("value")
                .and_then(|v| v.as_u64())
                .unwrap_or(0);
            outputs.insert("out".to_string(), Signal::from_bit(value as u8));
        }

        ComponentType::SevenSegment => {
            // Passthrough inputs for display
            for i in 0..7 {
                let key = format!("in_{}", i);
                let sig = inputs.get(&key).copied().unwrap_or(Signal::Low);
                outputs.insert(key, resolve(sig));
            }
        }

        ComponentType::HalfAdder => {
            let a = r(inputs.get("in_0").copied().unwrap_or(Signal::Low));
            let b = r(inputs.get("in_1").copied().unwrap_or(Signal::Low));
            outputs.insert("sum".to_string(), Signal::from_bit((a ^ b) & 1));
            outputs.insert("carry".to_string(), Signal::from_bit((a & b) & 1));
        }

        ComponentType::FullAdder => {
            let a = r(inputs.get("in_0").copied().unwrap_or(Signal::Low));
            let b = r(inputs.get("in_1").copied().unwrap_or(Signal::Low));
            let cin = r(inputs.get("in_2").copied().unwrap_or(Signal::Low));
            let axb = a ^ b;
            outputs.insert("sum".to_string(), Signal::from_bit((axb ^ cin) & 1));
            outputs.insert(
                "cout".to_string(),
                Signal::from_bit(((a & b) | (cin & axb)) & 1),
            );
        }

        ComponentType::Decoder => {
            let a = r(inputs.get("in_0").copied().unwrap_or(Signal::Low));
            let b = r(inputs.get("in_1").copied().unwrap_or(Signal::Low));
            let index = ((b as usize) << 1) | (a as usize);
            for i in 0..4 {
                outputs.insert(
                    format!("y{}", i),
                    Signal::from_bit(if i == index { 1 } else { 0 }),
                );
            }
        }

        ComponentType::BcdTo7Seg => {
            let d = r(inputs.get("in_0").copied().unwrap_or(Signal::Low));
            let c = r(inputs.get("in_1").copied().unwrap_or(Signal::Low));
            let b = r(inputs.get("in_2").copied().unwrap_or(Signal::Low));
            let a = r(inputs.get("in_3").copied().unwrap_or(Signal::Low));
            let val = ((d as usize) << 3) | ((c as usize) << 2) | ((b as usize) << 1) | (a as usize);

            // Active-high 7-segment patterns for hex digits 0-F: [A,B,C,D,E,F,G]
            const PATTERNS: [[u8; 7]; 16] = [
                [1, 1, 1, 1, 1, 1, 0], // 0
                [0, 1, 1, 0, 0, 0, 0], // 1
                [1, 1, 0, 1, 1, 0, 1], // 2
                [1, 1, 1, 1, 0, 0, 1], // 3
                [0, 1, 1, 0, 0, 1, 1], // 4
                [1, 0, 1, 1, 0, 1, 1], // 5
                [1, 0, 1, 1, 1, 1, 1], // 6
                [1, 1, 1, 0, 0, 0, 0], // 7
                [1, 1, 1, 1, 1, 1, 1], // 8
                [1, 1, 1, 1, 0, 1, 1], // 9
                [1, 1, 1, 0, 1, 1, 1], // A
                [0, 0, 1, 1, 1, 1, 1], // b
                [1, 0, 0, 1, 1, 1, 0], // C
                [0, 1, 1, 1, 1, 0, 1], // d
                [1, 0, 0, 1, 1, 1, 1], // E
                [1, 0, 0, 0, 1, 1, 1], // F
            ];

            let pattern = if val < 16 { PATTERNS[val] } else { [0; 7] };
            for i in 0..7 {
                outputs.insert(format!("out_{}", i), Signal::from_bit(pattern[i]));
            }
        }

        ComponentType::Mux2To1 => {
            let d0 = r(inputs.get("in_0").copied().unwrap_or(Signal::Low));
            let d1 = r(inputs.get("in_1").copied().unwrap_or(Signal::Low));
            let s = r(inputs.get("in_2").copied().unwrap_or(Signal::Low));
            outputs.insert("out".to_string(), Signal::from_bit(if s == 1 { d1 } else { d0 }));
        }

        ComponentType::Mux4To1 => {
            let d = [
                r(inputs.get("in_0").copied().unwrap_or(Signal::Low)),
                r(inputs.get("in_1").copied().unwrap_or(Signal::Low)),
                r(inputs.get("in_2").copied().unwrap_or(Signal::Low)),
                r(inputs.get("in_3").copied().unwrap_or(Signal::Low)),
            ];
            let s0 = r(inputs.get("in_4").copied().unwrap_or(Signal::Low));
            let s1 = r(inputs.get("in_5").copied().unwrap_or(Signal::Low));
            let sel = ((s1 as usize) << 1) | (s0 as usize);
            outputs.insert("out".to_string(), Signal::from_bit(d[sel]));
        }

        ComponentType::Demux1To4 => {
            let d = r(inputs.get("in_0").copied().unwrap_or(Signal::Low));
            let s0 = r(inputs.get("in_1").copied().unwrap_or(Signal::Low));
            let s1 = r(inputs.get("in_2").copied().unwrap_or(Signal::Low));
            let sel = ((s1 as usize) << 1) | (s0 as usize);
            for i in 0..4 {
                outputs.insert(
                    format!("y{}", i),
                    Signal::from_bit(if i == sel { d } else { 0 }),
                );
            }
        }

        ComponentType::SrLatch => {
            let s = r(inputs.get("in_0").copied().unwrap_or(Signal::Low));
            let rs = r(inputs.get("in_1").copied().unwrap_or(Signal::Low));
            let prev_q = properties
                .get("q")
                .and_then(|v| v.as_u64())
                .unwrap_or(0) as u8;
            let q = if s == 1 && rs == 0 {
                1
            } else if s == 0 && rs == 1 {
                0
            } else if s == 1 && rs == 1 {
                0 // invalid -> reset
            } else {
                prev_q
            };
            properties.insert("q".to_string(), serde_json::Value::from(q as u64));
            outputs.insert("q".to_string(), Signal::from_bit(q));
            outputs.insert("qn".to_string(), Signal::from_bit(if q == 1 { 0 } else { 1 }));
        }

        ComponentType::DFlipflop => {
            let d = r(inputs.get("in_0").copied().unwrap_or(Signal::Low));
            let clk = r(inputs.get("in_1").copied().unwrap_or(Signal::Low));
            let prev_clk = properties
                .get("prevClk")
                .and_then(|v| v.as_u64())
                .unwrap_or(0) as u8;
            let mut q = properties.get("q").and_then(|v| v.as_u64()).unwrap_or(0) as u8;
            let mut sampled_d = properties
                .get("sampledD")
                .and_then(|v| v.as_u64())
                .unwrap_or(d as u64) as u8;

            // Master-slave: sample while clock low
            if clk == 0 {
                sampled_d = d;
            }
            // Rising edge
            if prev_clk == 0 && clk == 1 {
                q = sampled_d;
            }

            properties.insert("prevClk".to_string(), serde_json::Value::from(clk as u64));
            properties.insert("sampledD".to_string(), serde_json::Value::from(sampled_d as u64));
            properties.insert("q".to_string(), serde_json::Value::from(q as u64));
            outputs.insert("q".to_string(), Signal::from_bit(q));
            outputs.insert("qn".to_string(), Signal::from_bit(if q == 1 { 0 } else { 1 }));
        }

        ComponentType::JkFlipflop => {
            let j = r(inputs.get("in_0").copied().unwrap_or(Signal::Low));
            let k = r(inputs.get("in_1").copied().unwrap_or(Signal::Low));
            let clk = r(inputs.get("in_2").copied().unwrap_or(Signal::Low));
            let prev_clk = properties
                .get("prevClk")
                .and_then(|v| v.as_u64())
                .unwrap_or(0) as u8;
            let mut q = properties.get("q").and_then(|v| v.as_u64()).unwrap_or(0) as u8;
            let mut sampled_j = properties
                .get("sampledJ")
                .and_then(|v| v.as_u64())
                .unwrap_or(j as u64) as u8;
            let mut sampled_k = properties
                .get("sampledK")
                .and_then(|v| v.as_u64())
                .unwrap_or(k as u64) as u8;

            // Master-slave: sample while clock low
            if clk == 0 {
                sampled_j = j;
                sampled_k = k;
            }
            // Rising edge
            if prev_clk == 0 && clk == 1 {
                q = match (sampled_j, sampled_k) {
                    (0, 0) => q,             // hold
                    (0, 1) => 0,             // reset
                    (1, 0) => 1,             // set
                    (1, 1) => if q == 1 { 0 } else { 1 }, // toggle
                    _ => q,
                };
            }

            properties.insert("prevClk".to_string(), serde_json::Value::from(clk as u64));
            properties.insert("sampledJ".to_string(), serde_json::Value::from(sampled_j as u64));
            properties.insert("sampledK".to_string(), serde_json::Value::from(sampled_k as u64));
            properties.insert("q".to_string(), serde_json::Value::from(q as u64));
            outputs.insert("q".to_string(), Signal::from_bit(q));
            outputs.insert("qn".to_string(), Signal::from_bit(if q == 1 { 0 } else { 1 }));
        }

        ComponentType::Comparator => {
            let a = r(inputs.get("in_0").copied().unwrap_or(Signal::Low));
            let b = r(inputs.get("in_1").copied().unwrap_or(Signal::Low));
            outputs.insert("gt".to_string(), Signal::from_bit(if a > b { 1 } else { 0 }));
            outputs.insert("eq".to_string(), Signal::from_bit(if a == b { 1 } else { 0 }));
            outputs.insert("lt".to_string(), Signal::from_bit(if a < b { 1 } else { 0 }));
        }

        ComponentType::Junction => {
            let in_signal = inputs
                .get("in")
                .copied()
                .or_else(|| inputs.values().next().copied())
                .unwrap_or(Signal::Low);
            let resolved = resolve(in_signal);
            for i in 0..5 {
                outputs.insert(format!("out_{}", i), resolved);
            }
        }

        ComponentType::Ic => {
            // IC logic handled by engine-level sub-engine evaluation
        }

        _ => {
            // Gates handled separately via evaluate_gate
        }
    }

    outputs
}
